interop-testing/src/test/java/io/grpc/testing/integration/CompressionTest.java

```diff
@@ -0,0 +1,272 @@
+  @AfterClass
+  public static void stopServer() {
+    stopStaticServer();
+  }
+
+  @Before
+  public void beforeTests() {
+    Fzip.INSTANCE.readIn = 0;
+    Fzip.INSTANCE.readOut = 0;
+    Fzip.INSTANCE.writeOut = 0;
+    Fzip.INSTANCE.writeIn = 0;
+  }
+
+  @After
+  public void afterTests() {
+    // Just check that the compressor was involved.
+    assertTrue(Fzip.INSTANCE.anyRead || Fzip.INSTANCE.anyWritten);
+  }
+
+  @Test
+  public void compresses() {
+    final SimpleRequest request = SimpleRequest.newBuilder()
+        .setResponseSize(314159)
+        .setResponseType(PayloadType.COMPRESSABLE)
+        .setPayload(Payload.newBuilder()
+            .setBody(ByteString.copyFrom(new byte[271828])))
+        .build();
+    final SimpleResponse goldenResponse = SimpleResponse.newBuilder()
+        .setPayload(Payload.newBuilder()
+            .setType(PayloadType.COMPRESSABLE)
+            .setBody(ByteString.copyFrom(new byte[314159])))
+        .build();
+
+    assertEquals(goldenResponse, blockingStub.unaryCall(request));
+    // Assert that compression took place
+    assertTrue(Fzip.INSTANCE.readIn < Fzip.INSTANCE.writeIn);
+  }
+
+  @Override
+  protected ManagedChannel createChannel() {
+    return ManagedChannelBuilder.forAddress("localhost", serverPort)
+        .decompressorRegistry(decompressors)
+        .compressorRegistry(compressors)
+        .intercept(new ClientInterceptor() {
+          @Override
+          public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+              MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+            final ClientCall<ReqT, RespT> call = next.newCall(method, callOptions);
+            return new ForwardingClientCall<ReqT, RespT>() {
+
+              @Override
+              protected ClientCall<ReqT, RespT> delegate() {
+                return call;
+              }
+
+              @Override
+              public void start(
+                  final ClientCall.Listener<RespT> responseListener, Metadata headers) {
+                ClientCall.Listener<RespT> listener = new ForwardingClientCallListener<RespT>() {
+
+                  @Override
+                  protected io.grpc.ClientCall.Listener<RespT> delegate() {
+                    return responseListener;
+                  }
+
+                  @Override
+                  public void onHeaders(Metadata headers) {
+                    super.onHeaders(headers);
+                    String encoding = headers.get(GrpcUtil.MESSAGE_ENCODING_KEY);
+                    assertEquals(encoding, Fzip.INSTANCE.getMessageEncoding());
+                  }
+                };
+                super.start(listener, headers);
+                setMessageCompression(true);
+              }
+            };
+          }
+        })
+        .usePlaintext(true)
+        .build();
+  }
+
+  private static final class Fzip implements Codec {
+    private static final Codec gzip = new Codec.Gzip();
```

Comment:
I am removing gzip, but it does serve a purpose:  it makes the input and output sizes different.  Its easier to tell that it actually did something.